### PR TITLE
CTCP-6012 Update guarantee balance format

### PR DESCRIPTION
**Pull request by karahanozturk**
https://jira.tools.tax.service.gov.uk/browse/CTCP-6012

### CODE REVIEW

File: app/viewmodels/package.scala [hunk 1]
```diff
@@ -695,24 +697,50 @@ package object viewmodels {
...
+          value.guaranteeType,
+          value.GuaranteeReference.flatMap(_.asString).takeSampleWithoutPeriod
+        ).slashSeparate
+      case "3" | "8" =>
+        Seq(
+          Some(value.sequenceNumber.toString),
+          Some(value.guaranteeType),
+          value.otherGuaranteeReference
+        ).flatten.slashSeparate
+    }
```
Comment (jackheslop96) [inline app/viewmodels/package.scala:738]: We probably don't need to worry about these cases. We wouldn't be getting to the point of them receiving an IE029 if their IE015 didn't have valid guarantee data. We can just do a one-size-fits-all here.

File: app/viewmodels/package.scala [hunk 2]
```diff
@@ -121,12 +121,14 @@ package object viewmodels {
...
     def take3(f: Seq[String] => String): String = takeN(3)(f)
 
-    def takeSample: String =
+    def takeSampleWithoutPeriod: String =
       if (value.length > 3) {
-        (Nil :+ value.head :+ ";...; " :+ value.last :+ ".").mkString
+        (Nil :+ value.head :+ ";...; " :+ value.last).mkString
       } else {
-        takeN(3)(_.mkString("; ")) + "."
+        takeN(3)(_.mkString("; "))
```
Comment (jackheslop96) [inline app/viewmodels/package.scala:128]: Could be `take3(_.mkString("; "))`

File: app/viewmodels/package.scala [hunk 3]
```diff
@@ -121,12 +121,14 @@ package object viewmodels {
     def take2(f: Seq[String] => String): String = takeN(2)(f)
     def take3(f: Seq[String] => String): String = takeN(3)(f)
 
-    def takeSample: String =
+    def takeSampleWithoutPeriod: String =
       if (value.length > 3) {
-        (Nil :+ value.head :+ ";...; " :+ value.last :+ ".").mkString
+        (Nil :+ value.head :+ ";...; " :+ value.last).mkString
```
Comment (jackheslop96) [inline app/viewmodels/package.scala:126]: Could we use `firstAndLast` here? Logic seems similar

File: app/viewmodels/package.scala [hunk 4]
```diff
@@ -695,24 +697,50 @@ package object viewmodels {
...
+
+    def normaliseOtherGuaranteeRefs: Seq[generated.GuaranteeType03] =
+      value
+        .groupBy(_.sequenceNumber)
+        .map {
+          (_, list) =>
+            val combinedOtherGuaranteeRefs = {
+              val refs = list.flatMap(_.otherGuaranteeReference)
+              if (refs.isEmpty) None
+              else Some(refs.takeSampleWithoutPeriod)
```
Comment (jackheslop96) [inline app/viewmodels/package.scala:710]: Very minor. Could use `Option.when(refs.nonEmpty)(refs.takeSampleWithoutPeriod)`

File: app/viewmodels/package.scala (hunk 2 above)
Comment (justega247) [inline app/viewmodels/package.scala:128]: Why are you using takeN(3) here, should the first check not cover it already?

Comment (justega247) [inline app/viewmodels/package.scala:128]: Do we need this takeN(3) method, can we not just use value.mkString(";  ")?

File: app/viewmodels/package.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (justega247) [reply app/viewmodels/package.scala:126]: I was thinking about the "/" example, and I am being particular about the ";...;" and not just "; "

File: app/viewmodels/package.scala (hunk 1 above)
Comment (justega247) [reply app/viewmodels/package.scala:738]: Just jumping on this, can you explain further @jackheslop96, I thought the case approach is to cater for the Guarantee reference field bit that might be present depending on the "type"?

Comment (jackheslop96) [reply app/viewmodels/package.scala:738]: @justega247 The rules in the DDNTA mean that certain data items will only be present for certain guarantee types. You can see this in our guarantee details front end navigation logic. So in production, by virtue of receiving a valid IE029, these cases become obsolete. If that makes sense?

We can review the rules tomorrow to see if I’ve got the wrong end of the stick!

File: app/viewmodels/package.scala [hunk 5]
```diff
@@ -695,24 +697,37 @@ package object viewmodels {
...
+        .groupBy(_.sequenceNumber)
+        .map {
+          (_, list) =>
+            val combinedOtherGuaranteeRefs = {
+              val refs = list.flatMap(_.otherGuaranteeReference)
+              Option.when(refs.nonEmpty)(refs.takeSampleWithoutPeriod)
+            }
+            list.head.copy(otherGuaranteeReference = combinedOtherGuaranteeRefs)
+        }
+        .toSeq
```
Comment (jackheslop96) [inline app/viewmodels/package.scala:713]: Just trying to work out what this is doing 😅 I'd have thought the sequence number is unique? Does that not make `.groupBy(_.sequenceNumber)` obsolete?

**Pull request merged**